### PR TITLE
TriConfig refactor

### PR DESCRIPTION
**Pull request by ndelangen**
## What I did

- Renamed `config.js` to `preview.js` (`config.js` still works)
- Renamed `addons.js` to `manager.js` (`addons.js` still works)
- Renamed `presets.js` to `main.js` (`presets.js` still works)
- I changed the docs to reflect the above changes.

- `webpack.config.js` is no longer required, this can now be placed in `main.js`, but it's call signature has changed to match presets.

- `main.js` is an actual preset! it can do everything a preset can do, plus it can expose an array of presets.
- this includes loading addons, setting webpack config for manager & preview, babel config etc.

- I removed the docs about `webpack.config.js` extend mode
- I changed the docs about theming to use the setConfig API

### CODE REVIEW

File: docs/src/pages/configurations/custom-webpack-config/index.md [hunk 1]
```diff
@@ -207,61 +205,36 @@ Nevertheless, edit `config` with care. Make sure to preserve the following confi
...
 ```js
-module.exports = async ({ config, mode }) => {
-  config.plugins.push(...)
-  return config;
-}
-```
-
-Finally, if your custom webpack config uses a loader that does not explicitly include specific file extensions via the `test` property, it is necessary to `exclude` the `.ejs` file extension from that loader.
-
-### Extend Mode (**Deprecated**)
```
Comment (tmeasday) [inline docs/src/pages/configurations/custom-webpack-config/index.md]: This is still supported in `webpack.config.js` though right?

File: examples/official-storybook/main.js [hunk 2]
```diff
@@ -0,0 +1,73 @@
+module.exports = {
```
Comment (tmeasday) [inline examples/official-storybook/main.js]: This looks great!

**Review by tmeasday — COMMENTED** [review]
Excited about this!

File: docs/src/pages/configurations/custom-webpack-config/index.md (hunk 1 above)
Comment (ndelangen) [reply docs/src/pages/configurations/custom-webpack-config/index.md]: yes, we support it.. but I think removing it from the docs, would be best.

Comment (tmeasday) [reply docs/src/pages/configurations/custom-webpack-config/index.md]: Yeah, probably, just wanted to check

**Review by shilman — COMMENTED** [review]
This looks really great!!! 🎉 

Stylistically my only suggestions are:

1) the `[].concat(x || [])` construct is a little hard to read. I'd suggest `if(x) y.push(x)` or `[x,y,z].filter(Boolean)`

2) for safety's sake, we may want to be more strict. logic like: "if there's a `manager.js`, use it. if there's also an `addons.js`, throw an error. if there's only an `addons.js` issue a deprecation warning (possibly we wait until 6.0 to deprecate?)." 

I think there's a lot more documentation to update, e.g. every doc that references `presets.js` should be updated. If we have both hanging around it will be confusing. Unless the idea is to make this a "soft launch" for now. Not sure how I feel about that.

And `MIGRATION.md` should be updated, explaining the change and what people need to do (for now, nothing).

Finally, this is a big change, so we should get feedback from @Hypnosphi @igor-dv @leoyli @CodeByAlex -- anybody who might have a strong opinion, rather than sliding it in now and then having the discussion later in the cycle.

**Comment by ndelangen** [comment]
If I hide `[].concat(x || [])` in a function, will that be ok? I'm used to writing it this way, and find it easier to track then `if(x){y.push(x)}` Though local mutation is fine, I like to avoid it.

I can't just array spread these, as strings are iterable and would result be spread into individual characters.

**Comment by ndelangen** [comment]
> for safety's sake, we may want to be more strict

What safety?

My proposal is to allow both, and possibly remove the old options at 6.0

**Comment by shilman** [comment]
> What safety?

If a user has both files (old and new) it will load both and lead to confusion. If the software forces the user into one paradigm or the other, it forces them to understand the change. Here's an example of that with CSF: https://github.com/storybookjs/storybook/blob/next/lib/core/src/client/preview/start.js#L364

**Comment by ndelangen** [comment]
@shilman should be good now

File: lib/core/src/server/common/custom-presets.js [hunk 3]
```diff
@@ -3,5 +3,10 @@ import serverRequire from '../utils/server-require';
 
 export default function loadCustomPresets({ configDir }) {
   const presets = serverRequire(path.resolve(configDir, 'presets'));
-  return presets || [];
+  const main = serverRequire(path.resolve(configDir, 'main'));
```
Comment (shilman) [inline lib/core/src/server/common/custom-presets.js]: Can we do the same sanity check here, to make `main` / `presets` mutually exclusive?

**Comment by shilman** [comment]
@ndelangen One more request above. Looking good!!!

**Comment by alolalo** [comment]
Hi Norbert & co,

I actually dropped an idea with Norbert that was taken into consideration; to include storybooks in storybooks.

So for feedback and future forwarding myself; I would only like to ask if that is taken into consideration since any storybook should be able to know if it is included and might need to use it's parent theming, or if it can set it's own, what other storybooks to include, what presets to overwrite, etc...

It's kind of specificity-spaghetti to me, noted that I'm not a front-end dev but a designer that does code. Hope this helps

**Review by shilman — APPROVED** [review]
Small is beautiful!!!

**Comment by ndelangen** [comment]
@alolalo yes it's in the works:
see https://github.com/storybookjs/storybook/pull/8720

**Pull request merged**